### PR TITLE
Moves most of FIRAuth initialization out of the calling thread.

### PR DESCRIPTION
**Pull request by XiangtianDai**
Also updates the sample app to verify the result.

### CODE REVIEW

File: AuthSamples/Sample/MainViewController.m [hunk 1]
```diff
@@ -493,6 +499,31 @@ @interface FIRApp (Internal)
 - (void)getTokenForcingRefresh:(BOOL)forceRefresh withCallback:(nonnull FIRTokenCallback)callback;
 @end
 
+@interface FIRAppAssociationRegistration (Deregistration)
```
Comment (protocol86) [inline AuthSamples/Sample/MainViewController.m]: please add @category comment here.

File: AuthSamples/Sample/MainViewController.m [hunk 2]
```diff
@@ -493,6 +499,31 @@ @interface FIRApp (Internal)
...
+
+/** @fn deregisteredObjectWithHost:key:
+    @brief Removes the object that was registered with a particular host and key, if one exists.
+    @param host The host object.
+    @param key The key to specify the registered object on the host.
+ */
++ (void)deregisterObjectWithHost:(id)host key:(NSString *)key;
+
```
Comment (protocol86) [inline AuthSamples/Sample/MainViewController.m]: no need to leave this space.

File: AuthSamples/Sample/MainViewController.m [hunk 3]
```diff
@@ -493,6 +499,31 @@ @interface FIRApp (Internal)
...
+@end
+
+@implementation FIRAppAssociationRegistration (Deregistration)
+
++ (void)deregisterObjectWithHost:(id)host key:(NSString *)key {
+  @synchronized(self) {
+    SEL dictKey = @selector(registeredObjectWithHost:key:creationBlock:);
+    NSMutableDictionary<NSString *, id> *objectsByKey = objc_getAssociatedObject(host, dictKey);
+    if (objectsByKey) {
```
Comment (protocol86) [inline AuthSamples/Sample/MainViewController.m]: may not matter in this case but we usually use arrayname.count for this test.

**Comment by paulb777** [comment]
Please check Mac build - 

From travis
```
 -> FirebaseCommunity (0.0.1)
    - ERROR | [OSX] [FirebaseCommunity/Auth] xcodebuild: Returned an unsuccessful exit code.
    - ERROR | [OSX] [FirebaseCommunity/Auth] xcodebuild:  /Users/travis/build/firebase/firebase-ios-sdk/Firebase/Auth/Source/FIRAuth.m:389:5: error: use of undeclared identifier 'UIApplication'
    - ERROR | [OSX] [FirebaseCommunity/Auth] xcodebuild:  /Users/travis/build/firebase/firebase-ios-sdk/Firebase/Auth/Source/FIRAuth.m:389:20: error: use of undeclared identifier 'application'; did you mean 'cApplication'?
    - NOTE  | [OSX] [FirebaseCommunity/Auth] xcodebuild:  /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.12.sdk/System/Library/Frameworks/CoreServices.framework/Frameworks/AE.framework/Headers/AERegistry.h:39:3: note: 'cApplication' declared here
    - ERROR | [OSX] [FirebaseCommunity/Auth] xcodebuild:  /Users/travis/build/firebase/firebase-ios-sdk/Firebase/Auth/Source/FIRAuth.m:389:35: error: unknown receiver 'UIApplication'; did you mean 'NSApplication'?
    - NOTE  | [OSX] [FirebaseCommunity/Auth] xcodebuild:  /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.12.sdk/System/Library/Frameworks/AppKit.framework/Headers/NSApplication.h:123:12: note: 'NSApplication' declared here
```

File: AuthSamples/Sample/MainViewController.m (hunk 1 above)
Comment (XiangtianDai) [inline AuthSamples/Sample/MainViewController.m]: Done

File: AuthSamples/Sample/MainViewController.m (hunk 2 above)
Comment (XiangtianDai) [inline AuthSamples/Sample/MainViewController.m]: Done

File: AuthSamples/Sample/MainViewController.m (hunk 3 above)
Comment (XiangtianDai) [inline AuthSamples/Sample/MainViewController.m]: Actually the entire check is unnecessary. Removed for conciseness.

**Pull request merged**